wren-core-py/src/extractor.rs

```diff
@@ -0,0 +1,306 @@
+        relationships: used_relationships,
+        metrics: mdl.metrics().to_vec(),
+        views: used_views,
+    })
+}
+
+fn extract_models(mdl: &Arc<WrenMDL>, used_datasets: &[String]) -> Vec<Arc<Model>> {
+    let mut used_set: HashSet<String> = used_datasets.iter().cloned().collect();
+    let mut stack: Vec<String> = used_datasets.to_vec();
+    while let Some(dataset_name) = stack.pop() {
+        if let Some(model) = mdl.get_model(&dataset_name) {
+            model
+                .columns
+                .iter()
+                .filter_map(|col| {
+                    col.relationship
+                        .as_ref()
+                        .and_then(|rel_name| mdl.get_relationship(rel_name))
+                })
+                .flat_map(|rel| rel.models.clone())
+                .filter(|related| used_set.insert(related.clone()))
+                .for_each(|related| stack.push(related));
+        }
+    }
+    mdl.models()
+        .iter()
+        .filter(|model| used_set.contains(model.name()))
+        .cloned()
+        .collect()
+}
+
+fn extract_views(
+    mdl: &Arc<WrenMDL>,
+    used_datasets: &[String],
+) -> (Vec<Arc<View>>, Vec<Arc<Model>>) {
+    let used_set: HashSet<&str> = used_datasets.iter().map(String::as_str).collect();
+    let stack: Vec<&str> = used_datasets.iter().map(String::as_str).collect();
+    let models = stack
+        .iter()
+        .filter_map(|&dataset_name| {
+            mdl.get_view(dataset_name).and_then(|view| {
+                resolve_used_table_names(mdl, view.statement.as_str())
+                    .ok()
+                    .map(|used_tables| extract_models(mdl, &used_tables))
+            })
+        })
+        .flatten()
+        .collect::<Vec<_>>();
+    let views = mdl
+        .views()
+        .iter()
+        .filter(|view| used_set.contains(view.name()))
+        .cloned()
+        .collect();
+
+    (views, models)
+}
+
+fn extract_relationships(
+    mdl: &Arc<WrenMDL>,
+    used_datasets: &[String],
+) -> Vec<Arc<Relationship>> {
+    let mut used_set: HashSet<String> = used_datasets.iter().cloned().collect();
+    let mut stack: Vec<String> = used_datasets.to_vec();
+    while let Some(dataset_name) = stack.pop() {
+        if let Some(relationship) = mdl.get_relationship(&dataset_name) {
+            for model in &relationship.models {
+                if used_set.insert(model.clone()) {
+                    stack.push(model.clone());
+                }
+            }
+        }
+    }
+    mdl.relationships()
+        .iter()
+        .filter(|rel| rel.models.iter().any(|model| used_set.contains(model)))
+        .cloned()
+        .collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::extractor::PyManifestExtractor;
+    use base64::prelude::BASE64_STANDARD;
+    use base64::Engine;
+    use rstest::{fixture, rstest};
+    use std::iter::Iterator;
+
+    #[fixture]
+    pub fn mdl_base64() -> String {
+        let mdl_json = r#"
```

Comment:
I prefer to use `ManifestBuilder` to build the case and then serialize it to the JSON if required. It's easier to maintain.

Reply:
Sure.